NullPointerException at game.ui.SpellView.updateSpells(SpellView.java:66)

NullPointerException

SpellHandler spellHandler =  WorldUtils.getWorld().orElse(null).getSystem(SpellHandler.class);

> ao-java/client/src/game/systems/render/world/TargetRenderingSystem.java:36: error: cannot find symbol
>             if (((GameScreen) game.getScreen()).getGUI().getSpellView().toCast.isPresent()) {
>                                                ^
>   symbol:   method getGUI()
>   location: class GameScreen
> ao-java/client/src/game/network/GameNotificationProcessor.java:104: error: cannot find symbol
>             ((GameScreen) game.getScreen()).getGUI().getInventory().updateUserInventory(0);
>                                            ^
>   symbol:   method getGUI()
>   location: class GameScreen
> ao-java/client/src/game/network/GameNotificationProcessor.java:163: error: cannot find symbol
>             final GUI gui = ((GameScreen) game.getScreen()).getGUI();
>                                                            ^
>   symbol:   method getGUI()
>   location: class GameScreen
> ao-java/client/src/game/ui/SpellView.java:60: error: cannot find symbol
>             ((GameScreen) game.getScreen()).getGUI().getConsole().addInfo("Haz click para lanzar el hechizo");
>                                            ^
>   symbol:   method getGUI()
>   location: class GameScreen
> Note: Some input files use unchecked or unsafe operations.
> Note: Recompile with -Xlint:unchecked for details.
> 4 errors

--- a/client/src/game/network/ClientResponseProcessor.java
+++ b/client/src/game/network/ClientResponseProcessor.java
@@ -24,6 +24,7 @@ public class ClientResponseProcessor extends BaseSystem implements IResponseProc
 
     private TimeSync timeSync;
 
+
     @Override
     public void processResponse(MovementResponse movementResponse) {
         MovementProcessorSystem.validateRequest(movementResponse.requestNumber, movementResponse.destination);

--- a/client/src/game/network/GameNotificationProcessor.java
+++ b/client/src/game/network/GameNotificationProcessor.java
@@ -5,7 +5,6 @@ import com.artemis.E;
 import com.artemis.Entity;
 import com.artemis.EntityEdit;
 import com.artemis.annotations.Wire;
-import com.badlogic.gdx.Game;
 import com.badlogic.gdx.Gdx;
 import com.esotericsoftware.minlog.Log;
 import entity.character.info.Inventory;
@@ -18,7 +17,6 @@ import game.screens.RoomScreen;
 import game.systems.camera.CameraShakeSystem;
 import game.ui.AOConsole;
 import game.ui.GUI;
-import game.utils.WorldUtils;
 import shared.network.interfaces.DefaultNotificationProcessor;
 import shared.network.inventory.InventoryUpdate;
 import shared.network.lobby.JoinRoomNotification;
@@ -37,6 +35,7 @@ public class GameNotificationProcessor extends DefaultNotificationProcessor {
     private WorldManager worldManager;
     private CameraShakeSystem cameraShakeSystem;
     private SoundsHandler soundsHandler;
+    private GUI gui;
 
     @Override
     public void processNotification(EntityUpdate entityUpdate) {
@@ -99,9 +98,7 @@ public class GameNotificationProcessor extends DefaultNotificationProcessor {
             }
         });
         // @todo fix
-        AOGame game = (AOGame) Gdx.app.getApplicationListener();
-        if (game.getScreen() instanceof GameScreen)
-            ((GameScreen) game.getScreen()).getGUI().getInventory().updateUserInventory(0);
+        gui.getInventory().updateUserInventory(0);
     }
 
     @Override
@@ -158,29 +155,26 @@ public class GameNotificationProcessor extends DefaultNotificationProcessor {
 
     @Override
     public void processNotification(ConsoleMessage consoleMessage) {
-        AOGame game = (AOGame) Gdx.app.getApplicationListener();
-        if (game.getScreen() instanceof GameScreen) {
-            final GUI gui = ((GameScreen) game.getScreen()).getGUI();
-            final AOConsole console = gui.getConsole();
-            final String message = consoleMessage.getMessage();
-            switch (consoleMessage.getKind()) {
-                case INFO:
-                    console.addInfo(message);
-                    break;
-                case ERROR:
-                    console.addError(message);
-                    break;
-                case COMBAT:
-                    console.addCombat(message);
-                    break;
-                case WARNING:
-                    console.addWarning(message);
-                    break;
-                default:
-                    console.addInfo(message);
-                    break;
-            }
+        final AOConsole console = gui.getConsole();
+        final String message = consoleMessage.getMessage();
+        switch (consoleMessage.getKind()) {
+            case INFO:
+                console.addInfo(message);
+                break;
+            case ERROR:
+                console.addError(message);
+                break;
+            case COMBAT:
+                console.addCombat(message);
+                break;
+            case WARNING:
+                console.addWarning(message);
+                break;
+            default:
+                console.addInfo(message);
+                break;
         }
+
     }
 
     @Override

--- a/client/src/game/screens/GameScreen.java
+++ b/client/src/game/screens/GameScreen.java
@@ -7,8 +7,6 @@ import com.artemis.WorldConfigurationBuilder;
 import com.artemis.managers.TagManager;
 import com.artemis.managers.UuidEntityManager;
 import com.badlogic.gdx.Gdx;
-import com.badlogic.gdx.InputMultiplexer;
-import com.badlogic.gdx.InputProcessor;
 import com.badlogic.gdx.ScreenAdapter;
 import com.badlogic.gdx.graphics.FPSLogger;
 import com.badlogic.gdx.graphics.OrthographicCamera;
@@ -54,8 +52,6 @@ public class GameScreen extends ScreenAdapter implements WorldScreen {
     private static final int DECORATION_PRIORITY = ENTITY_RENDER_PRIORITY - 2;
     public static World world;
     public static int player = -1;
-    private static GUI gui;
-    //private InputMultiplexer inputMultiplexer;
     private FPSLogger logger;
     private SpriteBatch spriteBatch;
     private WorldConfigurationBuilder worldConfigBuilder;
@@ -65,8 +61,6 @@ public class GameScreen extends ScreenAdapter implements WorldScreen {
         this.logger = new FPSLogger();
         long start = System.currentTimeMillis();
         initWorldConfiguration();
-        this.gui = new GUI();
-        gui.initialize();
         Gdx.app.log("Game screen initialization", "Elapsed time: " + TimeUnit.MILLISECONDS.toSeconds(Math.abs(System.currentTimeMillis() - start)));
     }
 
@@ -76,13 +70,8 @@ public class GameScreen extends ScreenAdapter implements WorldScreen {
 
     public static void setPlayer(int player) {
         GameScreen.player = player;
-        // @todo fix
-        gui.getInventory().updateUserInventory(0);
-        gui.getSpellView().updateSpells();
-    }
-
-    public GUI getGUI() {
-        return gui;
+        world.getSystem(GUI.class).getInventory().updateUserInventory(0);
+        world.getSystem(GUI.class).getSpellView().updateSpells();
     }
 
     public World getWorld() {
@@ -103,6 +92,8 @@ public class GameScreen extends ScreenAdapter implements WorldScreen {
                 .with(HIGH, new MovementAnimationSystem())
                 .with(HIGH, new IdleAnimationSystem())
                 .with(HIGH, new MovementSystem())
+                // GUI
+                .with(HIGH, new GUI())
                 // Camera
                 .with(HIGH, new CameraSystem(AOGame.GAME_SCREEN_ZOOM))
                 .with(HIGH, new CameraFocusSystem())
@@ -187,9 +178,6 @@ public class GameScreen extends ScreenAdapter implements WorldScreen {
     @Override
     public void render(float delta) {
         this.update(delta);
-        if (player >= 0) {
-            this.drawUI();
-        }
     }
 
     @Override
@@ -199,19 +187,15 @@ public class GameScreen extends ScreenAdapter implements WorldScreen {
         cameraSystem.camera.viewportHeight = cameraSystem.camera.viewportWidth * height / width;
         cameraSystem.camera.update();
 
-        gui.getStage().getViewport().update(width, height);
+        getWorld().getSystem(GUI.class).getStage().getViewport().update(width, height);
 
         getWorld().getSystem(LightRenderingSystem.class).resize(width, height);
-    }
-
-    private void drawUI() {
-        gui.draw();
     }
 
     @Override
     public void dispose() {
         world.getSystem(ClientSystem.class).stop();
-        gui.dispose();
+        world.getSystem(GUI.class).dispose();
         world.dispose();
     }
 

--- a/client/src/game/systems/render/world/TargetRenderingSystem.java
+++ b/client/src/game/systems/render/world/TargetRenderingSystem.java
@@ -8,8 +8,6 @@ import com.badlogic.gdx.Gdx;
 import com.badlogic.gdx.graphics.Color;
 import com.badlogic.gdx.graphics.Texture;
 import com.badlogic.gdx.graphics.g2d.SpriteBatch;
-import game.AOGame;
-import game.screens.GameScreen;
 import game.ui.GUI;
 import game.utils.Colors;
 import game.utils.WorldUtils;
@@ -23,6 +21,7 @@ import java.util.Optional;
 @Wire(injectInherited = true)
 public class TargetRenderingSystem extends RenderingSystem {
 
+    private GUI gui;
     private static Texture target = new Texture(Gdx.files.local("data/ui/images/target.png"));
 
     public TargetRenderingSystem(SpriteBatch batch) {
@@ -31,17 +30,14 @@ public class TargetRenderingSystem extends RenderingSystem {
 
     @Override
     protected void process(E e) {
-        AOGame game = (AOGame) Gdx.app.getApplicationListener();
-        if (game.getScreen() instanceof GameScreen) {
-            if (((GameScreen) game.getScreen()).getGUI().getSpellView().toCast.isPresent()) {
-                Optional<WorldPos> worldPos = WorldUtils.mouseToWorldPos();
-                if (worldPos.isPresent()) {
-                    Color prevColor = new Color(getBatch().getColor());
-                    getBatch().setColor(Colors.rgba(255, 255, 255, 0.5f));
-                    Pos2D mousePos = Util.toScreen(worldPos.get());
-                    getBatch().draw(target, mousePos.x, mousePos.y, Tile.TILE_PIXEL_WIDTH, Tile.TILE_PIXEL_HEIGHT);
-                    getBatch().setColor(prevColor);
-                }
+        if (gui.getSpellView().toCast.isPresent()) {
+            Optional<WorldPos> worldPos = WorldUtils.mouseToWorldPos();
+            if (worldPos.isPresent()) {
+                Color prevColor = new Color(getBatch().getColor());
+                getBatch().setColor(Colors.rgba(255, 255, 255, 0.5f));
+                Pos2D mousePos = Util.toScreen(worldPos.get());
+                getBatch().draw(target, mousePos.x, mousePos.y, Tile.TILE_PIXEL_WIDTH, Tile.TILE_PIXEL_HEIGHT);
+                getBatch().setColor(prevColor);
             }
         }
     }

--- a/client/src/game/ui/GUI.java
+++ b/client/src/game/ui/GUI.java
@@ -1,6 +1,8 @@
 package game.ui;
 
 import java.time.LocalDateTime;
+
+import com.artemis.BaseSystem;
 import com.badlogic.gdx.Gdx;
 import com.badlogic.gdx.Screen;
 import com.badlogic.gdx.graphics.OrthographicCamera;
@@ -17,9 +19,10 @@ import game.managers.AOInputProcessor;
 import game.screens.GameScreen;
 import game.ui.user.UserInformation;
 import game.utils.Skins;
+import net.mostlyoriginal.api.system.core.PassiveSystem;
 
 
-public class GUI implements Disposable {
+public class GUI extends BaseSystem implements Disposable  {
 
     //public static final int CONSOLE_TOP_BORDER = 16;
     //public static final int CONSOLE_LEFT_BORDER = 5;
@@ -36,6 +39,13 @@ public class GUI implements Disposable {
         this.userTable = new UserInformation();
         this.dialog = new DialogText();
         this.console = new AOConsole();
+    }
+
+    @Override
+    protected void processSystem() {
+        if (GameScreen.player >= 0) {
+            this.draw(world.getDelta());
+        }
     }
 
     public ActionBar getActionBar() {
@@ -83,6 +93,7 @@ public class GUI implements Disposable {
         pixmap.dispose();
     }
 
+    @Override
     public void initialize() {
         Skins.COMODORE_SKIN.getFont("flipped").setUseIntegerPositions(false);
         Skins.COMODORE_SKIN.getFont("flipped-shadow").setUseIntegerPositions(false);
@@ -145,8 +156,8 @@ public class GUI implements Disposable {
         return Gdx.graphics.getWidth();
     }
 
-    public void draw() {
-        stage.act(Gdx.graphics.getDeltaTime());
+    public void draw(float delta) {
+        stage.act(delta);
         stage.draw();
     }
 

--- a/client/src/game/ui/Inventory.java
+++ b/client/src/game/ui/Inventory.java
@@ -156,6 +156,9 @@ public class Inventory extends Window {
     }
 
     public void updateUserInventory(int base) {
+        if (base < 0) {
+            base = this.base;
+        }
         Item[] userItems = E(GameScreen.getPlayer()).getInventory().items;
         for (int i = 0; i < SIZE; i++) {
             Item item = base + i < userItems.length ? userItems[base + i] : null;

--- a/client/src/game/ui/SpellView.java
+++ b/client/src/game/ui/SpellView.java
@@ -55,11 +55,10 @@ public class SpellView extends Table {
     }
 
     private void changeCursor() {
-        AOGame game = (AOGame) Gdx.app.getApplicationListener();
-        if (game.getScreen() instanceof GameScreen) {
-            ((GameScreen) game.getScreen()).getGUI().getConsole().addInfo("Haz click para lanzar el hechizo");
-            Cursors.setCursor("select");
-        }
+        WorldUtils.getWorld().ifPresent(world -> {
+            world.getSystem(GUI.class).getConsole().addInfo("Haz click para lanzar el hechizo");
+        });
+        Cursors.setCursor("select");
     }
 
     public void updateSpells() {

--- a/client/src/game/ui/SpellView.java
+++ b/client/src/game/ui/SpellView.java
@@ -63,14 +63,15 @@ public class SpellView extends Table {
     }
 
     public void updateSpells() {
-        SpellHandler spellHandler =  WorldUtils.getWorld().orElse(null).getSystem(SpellHandler.class);
-        final Spell[] spells = spellHandler.getSpells();
-        Spell[] spellsToShow = new Spell[MAX_SPELLS];
-        System.arraycopy(spells, base, spellsToShow, 0, Math.min(MAX_SPELLS, spells.length));
-        for (int i = 0; i < MAX_SPELLS; i++) {
-            slots.get(i).setSpell(spellsToShow[i]);
-        }
-
+        WorldUtils.getWorld().ifPresent(world -> {
+            SpellHandler spellHandler =  world.getSystem(SpellHandler.class);
+            final Spell[] spells = spellHandler.getSpells();
+            Spell[] spellsToShow = new Spell[MAX_SPELLS];
+            System.arraycopy(spells, base, spellsToShow, 0, Math.min(MAX_SPELLS, spells.length));
+            for (int i = 0; i < MAX_SPELLS; i++) {
+                slots.get(i).setSpell(spellsToShow[i]);
+            }
+        });
     }
 
     private ImageButton createCastButton() {

--- a/client/src/game/utils/WorldUtils.java
+++ b/client/src/game/utils/WorldUtils.java
@@ -8,6 +8,7 @@ import com.badlogic.gdx.Screen;
 import com.badlogic.gdx.math.Vector3;
 import entity.character.states.Heading;
 import game.screens.GameScreen;
+import game.screens.ScreenManager;
 import game.screens.WorldScreen;
 import game.systems.camera.CameraSystem;
 import physics.AOPhysics;

--- a/desktop/build.gradle
+++ b/desktop/build.gradle
@@ -10,7 +10,6 @@ sourceSets {
     main.resources.srcDirs = ["assets/"]
 }
 
-
 dependencies {
     compile project(":client")
     compile "com.badlogicgames.gdx:gdx-backend-lwjgl3:$gdxVersion"

